### PR TITLE
ARROW-15802 [R] bindings for `lubridate::make_datetime()` and `lubridate::make_date()`

### PR DESCRIPTION
**Pull request by dragosmg**
- [x] bindings for `lubridate::make_datetime()` and `lubridate::make_date()`
- [x] bindings for `base::ISOdatetime()` and `base::ISOdate()`

### CODE REVIEW

File: r/R/dplyr-funcs-datetime.R [hunk 1]
```diff
@@ -187,6 +187,27 @@ register_bindings_datetime <- function() {
...
+                                             tz = NULL) {
+  x <- call_binding("paste", year, month, day, hour, min, sec, sep = "-")
+  # ParseTimestampStrptime currently ignores the timezone information (ARROW-12820).
+  # Stop if tz is provided.
+  if (is.character(tz)) {
+    arrow_not_supported("Time zone argument")
+  }
+
+  build_expr("strptime", x, options = list(format = "%Y-%m-%d-%H-%M-%S", unit = 0L))
+  })
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:210]: I wonder if it's possible (and advisable) to have `make_date` be defined as ~ `cast(make_datetime(year, month, day, hour = 0L, min = 0L, sec = 0))`? That will save a tiny bit of code (maybe it's not actually worth it?)

File: r/R/dplyr-funcs-datetime.R [hunk 2]
```diff
@@ -187,6 +187,27 @@ register_bindings_datetime <- function() {
...
+                                             hour = 0L,
+                                             min = 0L,
+                                             sec = 0,
+                                             tz = NULL) {
+  x <- call_binding("paste", year, month, day, hour, min, sec, sep = "-")
+  # ParseTimestampStrptime currently ignores the timezone information (ARROW-12820).
+  # Stop if tz is provided.
+  if (is.character(tz)) {
+    arrow_not_supported("Time zone argument")
+  }
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:207]: I wonder if we should set the default for tz to be `"UTC"` to match the signature of `make_datetime()` and then we can error here if it's not that value (and ignore it below). 

Or does that approach leave something else out?

**Review by jonkeane — CHANGES_REQUESTED** [review]
This is looking good. A few comments about structure + tests.

Looking at [the docs](https://lubridate.tidyverse.org/reference/make_datetime.html) for these, it looks like our implementation of `make_datetime()` could _also_ be mapped to `base:ISOdate()` and `base::ISOdatetime()`

File: r/tests/testthat/test-dplyr-funcs-datetime.R [hunk 3]
```diff
@@ -974,3 +974,50 @@ test_that("date() errors with unsupported inputs", {
...
+test_that("make_date & make_datetime", {
+  set.seed(12345)
+  test_df <- tibble(
+    year = sample(1969:2069, 12),
+    month = 1:12,
+    day = sample(1:28, 12, replace = TRUE),
+    hour = sample(0:23, 12, replace = TRUE),
+    min = sample(0:59, 12),
+    sec = sample(0:59, 12)
+  )
```
[the remark this answers is not in the document]
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-datetime.R:987]: I have a feeling we might need to wait to implement NA _progagation_ for `make_datetime()` and friends. I think the most elegant solution would rely on `strptime` returning an NA ([ARROW-15665](https://issues.apache.org/jira/browse/ARROW-15665))

[the remark this answers is not in the document]
Comment (dragosmg) [reply r/tests/testthat/test-dplyr-funcs-datetime.R]: Solved. Replaced the call to the `paste` (does not propagate `NA`s) binding with `str_c` (propagates `NA`s).

**Comment by dragosmg** [comment]
The are some outstanding issues, mostly having to do with the timezone argument:
* default values: 

  * for `ISOdate()` the default value is `"GMT"` which, in fact, is "UTC".
  * for `ISOdatetime()` the default value is `""` (current timezone) 
  * for `lubridate::make_datetime()` the default value is `"UTC"`
 
* the `strptime` binding does not support the `tz` argument - pointing to [ARROW-12820](https://issues.apache.org/jira/browse/ARROW-12820) - which is now solved. I have a feeling that issue doesn't actually mean we can now switch on `tz` support in `strptime`.

**Comment by jonkeane** [comment]
> The are some outstanding issues, mostly having to do with the timezone argument:
>
> 1. default values:
> for ISOdate() the default value is "GMT" which, in fact, is "UTC".
I think this is fine, we can make ours just be `"UTC"` it's most important to match behavior even if it means our signatures are slightly different.

> for ISOdatetime() the default value is "" (current timezone)
This is similar to our other issue with needing to add a TZ in arrow if we have "". Can we leave it as is and will that automagically become fixed when that other ticket is done? Speaking of: what _is_ the status of that other ticket?

> for lubridate::make_datetime() the default value is "UTC"
This is ok, right?


> 2. the strptime binding does not support the tz argument - pointing to [ARROW-12820](https://issues.apache.org/jira/browse/ARROW-12820) - which is now solved. I have a feeling that issue doesn't actually mean we can now switch on tz support in strptime.
Yes, though for this binding you can actually "just" specify the timezone after you do the `strptime`, no?

> For 1, the solution could be to set the default value for all tz args to "UTC".
Yup!

**Comment by dragosmg** [comment]
> Speaking of: what _is_ the status of that other ticket?

Still blocked by the Windows `tz` database availability issue #12536.

**Comment by dragosmg** [comment]
> This is similar to our other issue with needing to add a TZ in arrow if we have "". Can we leave it as is and will that automagically become fixed when that other ticket is done?

Not sure that will _automagically_ be solved since `strptime` will still not accept a `tz` argument (and that is the barrier).

**Comment by dragosmg** [comment]
> Yes, though for this binding you can actually "just" specify the timezone after you do the `strptime`, no?

Not sure I follow, once `x` is created, I don't think I can change the value of `x$type$timezone` (it seems to be locked).

File: r/R/dplyr-funcs-datetime.R [hunk 4]
```diff
@@ -187,6 +187,51 @@ register_bindings_datetime <- function() {
...
+    x <- call_binding("str_c", year, month, day, hour, min, sec, sep = "-")
+    build_expr("strptime", x, options = list(format = "%Y-%m-%d-%H-%M-%S", unit = 0L))
+  })
+  register_binding("make_date", function(year = 1970L, month = 1L, day = 1L) {
+    x <- call_binding("make_datetime", year, month, day)
+    build_expr("cast", x, options = cast_options(to_type = date32()))
+  })
+  register_binding("ISOdatetime", function(year,
+                                           month,
+                                           day, hour,
```
Comment (jonkeane) [inline r/R/dplyr-funcs-datetime.R:213]:
```suggestion
                                           day, 
                                           hour,
```

Tiny nit: for alignment?

**Review by jonkeane — APPROVED** [review]
This looks good. Thanks!

**Pull request closed without merging**